### ISSUE TITLE
Unable to add Python objects to the QueryScope

### ISSUE THREAD

**Issue by chipkent** [deephaven/deephaven-core#846]
Python objects can not be added to the QueryScope manually.  The same objects can be added automatically at the top level without problems.  I suspect this is a problem with assigning Python objects to `java.lang.Object` as illustrated below.

```
from deephaven import QueryScope
from deephaven.TableTools import emptyTable

class MyClass:
    def __init__(self):
        pass

    def do_it(self):
        return 3


def in_func(t):
    mc2 = MyClass()
    #This fails with an error
    QueryScope.addParam("mc2", mc2)
    return t.update("X = mc2.do_it()")

t = emptyTable(3)

# this works fine
mc1 = MyClass()
t1 = t.update("X = mc1.do_it()")

# This results in an error
t2 = in_func(t)

```

![image](https://user-images.githubusercontent.com/5250374/125860327-287b8629-ef51-4a48-a92b-9ac6815da98f.png)

I suspect the problem is an inability to assign a Python object to a `java.lang.Object`.  This can be seen here:

```

import jpy

class MyClass:
    def __init__(self):
        pass

    def do_it(self):
        return 3

mc1 = MyClass()

a = jpy.array("java.lang.Object", 3)
a[0] = mc1
```
![image](https://user-images.githubusercontent.com/5250374/125860568-f1c82992-655d-4538-85bc-4b084f6cc11e.png)

**Comment by jmao-denver** [comment]
The exception is thrown by JPY. Basically JPY doesn't know how to convert a Python object (other than primitive values) to a corresponding java.lang.Object object based on the signatures of the QueryScope.addParam() which is generic method. 

We can undoubtedly change this JPY behavior to always wrap a non-primitive Python object into a org.jpy.PyObject object. This will allow the JVM to access any Python object not just the top level ones (in the main globals). The danger here is that a Python object may go out of scope and the JVM has no idea of that happening. e.g. an object created inside a function. And even if the JVM holds on to a function-scope object by incrementing the ref count, there is a high change the Python interpreter will find it a fatal error when it tries to release all the auto variables upon the function exist.

To summarize:
1. It should be possible to support the use case as requested by changing JPY's behavior when it encounters a java.lang.Object type parameter on calling a Java method.
2. there are important questions to be answered around the risks of putting a non-main-module-level object into the global QueryScope, and come up with solutions to make it clearly understood by the users and thus make this new feature safe to use.

**Comment by rcaudy** [comment]
It looks to me like we have two choices here:
1. For now, there's an obvious workaround: Document that users must raise variables to the main module (i.e. return them) before trying to add them to the `QueryScope`.
2. Meanwhile, we can add either a function that wraps query scope addition (or a full wrapper `QueryScope` implementation) to make the Python integration better and allow this workflow (wrap a Python object in a `PyObject`).

We should *not* make it so that JPY chooses to always convert Python objects to `PyObject` to call methods that take `java.lang.Object`.

**Comment by jmao-denver** [comment]
Prompted by the call with @rcaudy and @cpwright, I have dug a little deeper and would go so far to suggest that we get rid of of QueryScope for Python and instead just teach how the user can use Python globals() to make any Python objects available for query evaluation. For example, in @chip's example, the problematic function can be rewritten as:
```
def in_func(t):
    mc2 = MyClass()
    globals()["mc2"]= mc2
    return t.update("X = mc2.do_it()")
```
Of course, this is based on my limited knowledge of what QueryScope was designed for and how it is being used in the fields. But short of making it a true DH construct other than a wrapper around Python globals, teaching the users that DH query engine only makes use of Python globals and the users are responsible for the objects they create and want to be accessed inside DH could be a cleaner solution to the confusion and frustration highlighted in this issue.

**Comment by rcaudy** [comment]
That's not a terrible suggestion for the status quo. It's a bit ugly, but it's very clear how to get what you want and what the implications are (i.e. that making something visible to DH means putting it in Python globals).

That said, there *is* value in adding a wrapper if we might eventually change the implementation away from using Python globals and instead require users to explicitly add scope variables. I'm not sure how likely that is, however.

**Comment by chipkent** [comment]
Let me propose a concrete use case to discuss.  All of this is in python.  This is pseudocode, so some module or function names may be wrong.

```
from deephaven.DBTimeUtils import convertDateTime

def func(t):
    aa = convertDateTime("2021-01-01T08:23:01.23 NY")
    <*** ADD aa TO SCOPE ***>
    return t.where("X > aa")

t = ...
t1 = func(t)
```

In this example, `aa` is a Java object wrapped by jpy.  Given the proposal, how should `aa` be added to the scope of the query string?

1. Add the Java object to the query scope using python `globals()`?
2. Add the Java object to the query scope using `QueryScope`?

- Would `globals()` function in this case? -- I'm guessing so.
- Does `globals()` introduce a bunch of inefficiencies through Python which would not be present if the Java variable was registered via `QueryScope`? -- Maybe, I'm not sure.
- If there are problems in the above cases, then we have a problem with users needing to know what kind of object they have and what way to register it efficiently.

**Comment by rcaudy** [comment]
>Does globals() introduce a bunch of inefficiencies through Python which would not be present if the Java variable was registered via QueryScope? -- Maybe, I'm not sure.

No. That's literally what the `QueryScope` is in a Python session.

**Comment by chipkent** [comment]
That makes the error we are seeing even more weird.  So we have python->java->python, and the java types in the middle are just blocking the put into the python globals?

**Comment by rcaudy** [comment]
Yes. Exactly that.

**Comment by chipkent** [comment]
I guess that is easier to fix.

**Comment by rcaudy** [comment]
Hence the workaround Jianfeng published, to just use `globals()`.

**Comment by chipkent** [comment]
I want to get the thoughts of the DevRel team before plowing ahead.  If we go this way, it seems like we have two bits of work:
1) Remove `QueryScope` from the python library.
2) Find and fix all python `QueryScope` documentation.

**Comment by hythloda** [comment]
The symmetry of having QueryScope in python is nice, but I don't mind the idea of globals. In just my limited use of them (last 30 minutes) I have not found a break yet with globals.

**Comment by jjbrosnan** [comment]
I haven't been able to find other cases to break things in the last hour or so.  I may have luck with more time.  I wouldn't mind using globals to avoid this QueryScope issue

**Comment by chipkent** [comment]
I just discovered #883 .  It is somewhat related to this problem.

**Comment by chipkent** [comment]
The documentation side of these fixes can be found in:
https://github.com/deephaven/deephaven.io/pull/469
